### PR TITLE
CyclicLR memory leak fix

### PR DESCRIPTION
**Pull request by kongzii**
Hi, we noticed in our team that by using CyclicLR, there is a problem with memory clearance on GPU (probably it will be the case without the GPU as well, but that was our use case) After initializing CyclicLR, GPU memory is not cleared even after the model, optimizer and scheduler are out of scope (e.g. reference count is zero). This is because `__init__` method inside `CyclicLR` creates reference to its own methods and it will not get removed until `gc.collect()` is called manually. This is a problem if people want to test multiple models in one run of a script, after testing the first model, second one will fail on `CUDA out of memory error` because the first one is not cleared from the memory.

I propose a simple fix by using `weakref`, similarly as in `_LRScheduler` base class, but if you have any comments I am happy to change it. 

Here is the code to reproduce the bug:

```
import torch
import weakref
from transformers import DetrForObjectDetection

class X:
    def __init__(self, optimizer):
        self.optimizer = optimizer

        # Will cause cyclic reference.
        self.func = self.dummy

        # Will work as expected, memory cleared after instance count is zero.
        # self.func = weakref.WeakMethod(self.dummy)

    def dummy(self, x):
        return 1.

def test():
    model = DetrForObjectDetection.from_pretrained('facebook/detr-resnet-50')
    model.to('cuda')
    optimizer = torch.optim.Adam(model.parameters())
    x = X(optimizer)

test()
print(f'{torch.cuda.memory_reserved()}, {torch.cuda.memory_allocated()}')  # Should print (<some memory>, 0), but with cyclic reference, it will print (<some memory>, <some memory>).
```

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -1214,17 +1214,19 @@ def __init__(self,
         self.gamma = gamma
 
         if scale_fn is None:
+            self.scale_fn_custom = None
```
Comment (albanD) [inline torch/optim/lr_scheduler.py:1217]: Can you name these `self._scale_fn_custom` since it should be private (same for `_scale_fn_ref`.

**Review by albanD — COMMENTED** [review]
That looks good!
Could you add a small test to make sure we don't regress this in the future? You can do so by
- Disable the gc via gc.disable()
- Have a function that create such scheduler and return a weak_ref to it
- Make sure the weak ref is empty as soon as you exit the function even without a gc.collect

**Comment by kongzii** [comment]
@albanD Sorry for the late reply, both test and renaming to private variables are done. One can easily verify that the test will fail if you include something like this inside CyclicLR's __init__ method:

```python
self.test = self._triangular_scale_fn
```

**Review by albanD — APPROVED** [review]
Awesome! Thanks for the updated!

**Comment by albanD** [comment]
Could you please fix the lint issues?

**Comment by kongzii** [comment]
> Could you please fix the lint issues?

@albanD Sorry! When I execute `lintrunner` locally I have no errors, probably a bad setup. Fixed.

**Review by albanD — APPROVED** [review]

**Comment by kongzii** [comment]
Seems I do not have permission to add any label?

**Comment by albanD** [comment]
Added!
In the future, you can ask the bot to do it for you as needed:
@pytorchbot help

**Pull request closed without merging**